### PR TITLE
Get instance in Openstack compute plugin

### PR DESCRIPTION
**Pull request by llvieira**
[GitLab task.](https://git.lsd.ufcg.edu.br/chicog/fogbow/issues/581)

Please, review this new pull request that implements getInstance() in Openstack compute plugin. All tests are included.

### CODE REVIEW

File: src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java [hunk 1]
```diff
@@ -361,15 +365,54 @@ protected String doGetRequest(String endpoint, Token localToken) throws RequestE
...
-        return null;
+        LOGGER.info("Getting instance " + instanceId + " with token " + localToken);
+
+        String tenantId = getTenantId(localToken);
+        String requestEndpoint = getComputeEndpoint(tenantId, SERVERS + "/" + instanceId);
+
+        String jsonResponse = doGetRequest(requestEndpoint, localToken);
+
+        LOGGER.debug("Getting instance from json: " + jsonResponse);
+        return getInstanceFromJson(jsonResponse);
```
Comment (paulofelipefeitosa) [inline src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java]: Please, do not return a method return value, it is bad for debug.

File: src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java [hunk 2]
```diff
@@ -361,15 +365,54 @@ protected String doGetRequest(String endpoint, Token localToken) throws RequestE
...
-        return null;
+    private ComputeOrderInstance getInstanceFromJson(String jsonResponse) throws RequestException {
+        try {
+            JSONObject rootServer = new JSONObject(jsonResponse);
+            JSONObject serverJson = rootServer.getJSONObject(SERVER_JSON_FIELD);
+
+            String id = serverJson.getString(ID_JSON_FIELD);
+            InstanceState state = getInstanceState(serverJson.getString(STATUS_JSON_FIELD));
+
+            return new ComputeOrderInstance(id, state);
```
Comment (paulofelipefeitosa) [inline src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java]: Please, do not return a method return value, it is bad for debug.

File: src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java [hunk 3]
```diff
@@ -361,15 +365,54 @@ protected String doGetRequest(String endpoint, Token localToken) throws RequestE
...
     }
 
-    private List<ComputeOrderInstance> getInstancesFromJson(String json) {
+    /**
+     * This method will map Openstack instance status to Fogbow instance status.
+     *
+     * @param instanceStatus status from JSON.
+     * @return {@link InstanceState}
+     */
+    private InstanceState getInstanceState(String instanceStatus) {
```
Comment (paulofelipefeitosa) [inline src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java]: **OpenStack** server creation status transition:
- **BUILD** (The server has not yet finished the original build process).
- **ACTIVE** (The server is active).
- **ERROR** (The server is in error).

So, if think that the InstanceState should be **INACTIVE** only when the openstack server is in the **BUILD** state. The default state I think that can be **FAILED**.

File: src/main/java/org/fogbowcloud/manager/core/models/orders/instances/ComputeOrderInstance.java [hunk 4]
```diff
@@ -31,6 +32,11 @@ public ComputeOrderInstance(String id, String hostName, int vCPU, int memory, In
         this.sshExtraPorts = sshExtraPorts;
     }
 
+    // TODO: This method was created to make getInstance() more easy to implement
```
Comment (thiagomanel) [inline src/main/java/org/fogbowcloud/manager/core/models/orders/instances/ComputeOrderInstance.java]: getInstance from which class?

File: src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java (hunk 3 above)
Comment (thiagomanel) [reply src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java]: I'm not sure I get the context (from both the code and the comment). @fgan1 please, take a look

File: src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java [hunk 5]
```diff
@@ -361,15 +365,54 @@ protected String doGetRequest(String endpoint, Token localToken) throws RequestE
...
+
+            case ERROR_STATUS:
+                return InstanceState.FAILED;
+
+            default:
+                return InstanceState.INACTIVE;
+        }
+    }
+
+    // TODO: Is it necessary?
```
Comment (thiagomanel) [inline src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java]: was anyone using it on the old code base?

File: src/main/java/org/fogbowcloud/manager/core/models/orders/instances/ComputeOrderInstance.java (hunk 4 above)
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/models/orders/instances/ComputeOrderInstance.java]: getInstance is to get a Virtual Machine from Openstack. It's not related to class' instances.

File: src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java (hunk 2 above)
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java]: Fixed!

File: src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java (hunk 5 above)
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java]: I don't think so.

File: src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java (hunk 1 above)
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java]: Fixed!

**Pull request merged**